ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/ChainDB/Impl/Types.hs

```diff
@@ -512,51 +547,95 @@ addReprocessLoEBlocks
 -- callback.
 getChainSelMessage
-  :: (IOLike m, HasHeader blk)
+  :: (HasHeader blk, IOLike m)
   => Tracer m (TraceChainSelStarvationEvent blk)
   -> StrictTVar m ChainSelStarvation
   -> ChainSelQueue m blk
   -> m (ChainSelMessage m blk)
-getChainSelMessage starvationTracer starvationVar (ChainSelQueue queue) = do
-  -- NOTE: The test of emptiness and the blocking read are in different STM
-  -- transactions on purpose.
-  whenM (atomically $ isEmptyTBQueue queue) $ do
-    writeTVarIO starvationVar ChainSelStarvationOngoing
-    traceWith starvationTracer . ChainSelStarvationStarted =<< getMonotonicTime
-  message <- atomically $ readTBQueue queue
-  -- If there was a starvation ongoing, we need to report that it is done.
-  whenM ((== ChainSelStarvationOngoing) <$> readTVarIO starvationVar) $
-    case message of
-      ChainSelAddBlock BlockToAdd {blockToAdd} -> do
-        time <- getMonotonicTime
-        traceWith starvationTracer $ ChainSelStarvationEnded time $ blockRealPoint blockToAdd
-        writeTVarIO starvationVar $ ChainSelStarvationEndedAt time
-      ChainSelReprocessLoEBlocks -> pure ()
-  return message
+getChainSelMessage starvationTracer starvationVar queue = go
   where
+    go = do
+      (reprocessLoEBlocks, chainSelQueue) <- atomically readBoth
+      case reprocessLoEBlocks of
+        True -> do
+          writeTVarIO varChainSelReprocessLoEBlocks False
+          return ChainSelReprocessLoEBlocks
+        False ->
+          case Map.minView chainSelQueue of
+            Just (blockToAdd, chainSelQueue') -> do
+              writeTVarIO varChainSelQueue chainSelQueue'
+              terminateStarvationMeasure blockToAdd
+              return $ ChainSelAddBlock blockToAdd
+            Nothing -> do
+              startStarvationMeasure
+              void $ atomically $ blockUntilChanged (second Map.null) (False, True) readBoth
+              go
+    ChainSelQueue {varChainSelQueue, varChainSelReprocessLoEBlocks} = queue
     writeTVarIO v x = atomically $ writeTVar v x
+    swapTVarIO v x = atomically $ swapTVar v x
+    readBoth = (,) <$> readTVar varChainSelReprocessLoEBlocks <*> readTVar varChainSelQueue
+
+    startStarvationMeasure = do
+      prevStarvation <- swapTVarIO starvationVar ChainSelStarvationOngoing
+      when (prevStarvation /= ChainSelStarvationOngoing) $
+        traceWith starvationTracer . ChainSelStarvationStarted =<< getMonotonicTime
+
+    terminateStarvationMeasure BlockToAdd{blockToAdd=block} = do
+      prevStarvation <- readTVarIO starvationVar
+      when (prevStarvation == ChainSelStarvationOngoing) $ do
+        tf <- getMonotonicTime
+        traceWith starvationTracer (ChainSelStarvationEnded tf $ blockRealPoint block)
+        writeTVarIO starvationVar (ChainSelStarvationEndedAt tf)
 
 -- | Flush the 'ChainSelQueue' queue and notify the waiting threads.
 --
+-- REVIEW: What about all the threads that are waiting to write in the queue and
+-- will write after the flush?!
```

Comment:
Indeed, there could be short time after the queue is flushed, but before the ChainDB is closed (and hence nothing new can be added to the ChainDB). I think this already exists on `main`.

I raised this on the IOG Slack, but I don't think we need to do anything about it in this PR.

Reply:
In that case I'd say: let's add a note and remove the `REVIEW` item.